pkg/mappings.bzl

```diff
@@ -0,0 +1,403 @@
+# Copyright 2021 The Bazel Authors. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Package creation helper mapping rules.
+
+This module declares Provider interfaces and rules for specifying the contents
+of packages in a package-type-agnostic way.  The main rules supported here are
+the following:
+
+- `pkg_files` describes destinations for rule outputs
+- `pkg_mkdirs` describes directory structures
+
+Rules that actually make use of the outputs of the above rules are not specified
+here.  TODO(nacl): implement one.
+"""
+
+load("@bazel_skylib//lib:paths.bzl", "paths")
+load("//:providers.bzl", "PackageDirsInfo", "PackageFilesInfo")
+
+_PKGFILEGROUP_STRIP_ALL = "."
+
+def _sp_files_only():
+    return _PKGFILEGROUP_STRIP_ALL
+
+def _sp_from_pkg(path = ""):
+    if path.startswith("/"):
+        return path[1:]
+    else:
+        return path
+
+def _sp_from_root(path = ""):
+    if path.startswith("/"):
+        return path
+    else:
+        return "/" + path
+
+strip_prefix = struct(
+    _doc = """pkg_files `strip_prefix` helper.  Instructs `pkg_files` what to do with directory prefixes of files.
+
+    Each member is a function that equates to:
+
+    - `files_only()`: strip all directory components from all paths
+
+    - `from_pkg(path)`: strip all directory components up to the current
+      package, plus what's in `path`, if provided.
+
+    - `from_root(path)`: strip beginning from the file's WORKSPACE root plus
+      what's in `path`, if provided.
+""",
+    files_only = _sp_files_only,
+    from_pkg = _sp_from_pkg,
+    from_root = _sp_from_root,
+)
+
+####
+# Internal helpers
+####
+
+def _validate_attr(attr):
+    # TODO(nacl): this needs to be rethought.  There could, for example, be
```

Comment:
There is not much winning here. What starlark needs is attr.string_dict_dict , so we can have a dict of dicts.   What would be better is if you could declare attributes to be of a provider type, and then we get type safe complex args.   How about an idea like this
```
def file_attributes(owner=None, group=None, mode=None, **kwargs):
  ret = {}
  if owner:
    ret['owner'] = owner
  if group:
    ret['group'] = group
  ret['mode'] = mode or 0o444
  for key in kwargs:
    ret[key] = kwargs[key]
  return json.encode(ret)

def _inspect_impl(ctx):
  print('==== attributes')
  for plat in ctx.attr.attributes:
    print(plat, json.decode(ctx.attr.attributes[plat]))
```
and in BUILD
```
inspect(
    name = "i",
    target = ":tfiles",
    attributes = {
        'unix': file_attributes(mode=0o777, foo='bar'),
        'windows': file_attributes(mode=0o444, foo='baz'),
    }
)
```

Reply:
OK. We can do it later.
What I need to do in the interim is explain my stance that the `file_attributes` macro approach should be a standard idiom in a doc and pass it around for people to think about. This keeps coming up in other rules too.

FWIW, what I really want is
```
rule(... attrs = {"attributes": attr.list_of_providers(providers=[a, b, c])})
```
Such that get real type safety.  You can't just put anything there, it must be a provider instantiation.
I expect there will be resistance. :-)